File: tools/packfile.py
#!/usr/bin/env python3
"""Read a shipped pack: its manifest, and the modules that manifest names.

A pack is not one file. It is a `manifest.json` plus a set of content-addressed
module files in a directory they all share, and a section may be split across
two of them -- its structure in `core`, the game's own language in `names` --
under the same key in both.

The modules holding a language are named per language, so reading a pack means
choosing one. Every reader here defaults to `DEFAULT_LOCALE`, which every pack
ships and which is what the app shows unless somebody asks otherwise.

Every tool that reads a pack reads it through here. Four of them used to open
one gzipped document directly, and four re-implementations of "join the modules
back together" would be four chances to join them differently.
"""

from __future__ import annotations

import gzip
import json
from pathlib import Path
from typing import Any

SITE = Path(__file__).resolve().parent.parent / "site"
"""The document root every module path in a manifest is relative to."""

DEFAULT_LOCALE = "enUS"
"""Which language a reader gets when it does not ask for one.

Declared again from `pack.derive.locales`, because a tool must not have to put
the build package on its path to read a pack that is already on disk.
`check_locale_declaration` reconciles the two.
"""


def manifest_of(pack_dir: Path) -> dict[str, Any]:
    """One pack's manifest.

    Raises:
        FileNotFoundError: there is no pack there. Raised rather than returning
            an empty manifest, which would read downstream as a pack that ships
            nothing -- a different and much quieter kind of wrong.
    """
    path = pack_dir / "manifest.json"
    if not path.exists():
        raise FileNotFoundError(f"{path} does not exist")
    loaded = json.loads(path.read_text(encoding="utf-8"))
    assert isinstance(loaded, dict)
    return loaded


def module(file: str) -> dict[str, Any]:
    """One module file, decoded."""
    with gzip.open(SITE / file, "rt", encoding="utf-8") as handle:
        loaded = json.load(handle)
    assert isinstance(loaded, dict)
    return loaded


def entries(manifest: dict[str, Any],
            locale: str = DEFAULT_LOCALE) -> dict[str, dict[str, Any]]:
    """Which module file a pack reads for each module, in one language.

    The structure modules and the chosen language's, together: that IS a pack
    to a reader, and every consumer wanting one asks for it the same way.

    Args:
        manifest: the pack's manifest.
        locale: the language to read. A pack that does not carry it falls back
            to the default one, which every pack ships.

    Returns:
        Module name -> its manifest entry.

    Raises:
        ValueError: the manifest names no languages at all. That is a pack from
            before the language became an axis of the artifact, and reading it
            as though it merely had none would silently drop every name and
            every description in it.
    """
    locales = manifest.get("locales", {})
    if not locales:
        raise ValueError(f"{manifest.get('pack', 'this pack')} names no languages; "
                         f"it predates the language axis and has to be rebuilt")
    return {**manifest.get("modules", {}),
            **(locales.get(locale) or locales.get(DEFAULT_LOCALE, {}))}


def load(pack_dir: Path, *, want: tuple[str, ...] = (),
         locale: str = DEFAULT_LOCALE) -> dict[str, Any]:
    """One pack as the single document its modules add up to.

    Args:
        pack_dir: the pack's own directory, holding its manifest.
        want: which modules to read, by name. Empty means all of them -- but a
            caller after asset paths has no use for prose, and `text` is a
            quarter of the pack, so naming what you need is worth it.
        locale: which language's names and prose to read.

    Returns:
        Every section by name, with `meta` from the manifest. A section split
        across modules comes back joined, which is what makes a reader here
        indifferent to which module a column ended up in.
    """
    manifest = manifest_of(pack_dir)
    pack: dict[str, Any] = {"meta": manifest.get("meta", {})}
    for name, entry in entries(manifest, locale).items():
        if want and name not in want:
            continue
        for section, columns in module(entry["file"]).items():
            held = pack.get(section)
            if isinstance(held, dict) and isinstance(columns, dict):
                held.update(columns)
            else:
                pack[section] = columns
    return pack


def sizes(pack_dir: Path, locale: str = DEFAULT_LOCALE) -> dict[str, int]:
    """What each of a pack's modules costs, by module name.

    Off the manifest rather than off the files, so it answers without reading
    a megabyte -- which is the reason the sizes are written down there.

    One language's, because that is what a reader downloads: adding every
    language up would describe a pack nobody fetches.
    """
    return {name: entry["bytes"]
            for name, entry in entries(manifest_of(pack_dir), locale).items()}


def files(manifest: dict[str, Any]) -> list[str]:
    """Every module file a pack names, in every language it ships.

    What a deploy has to carry, as against what one reader fetches: a pack is
    only complete when the file behind each of these is there.
    """
    named = [entry["file"] for entry in manifest.get("modules", {}).values()]
    for spoken in manifest.get("locales", {}).values():
        named += [entry["file"] for entry in spoken.values()]
    return named

File: tools/packfile.ts
/**
 * @file Reading a shipped pack off disk, for the tools that run outside a browser.
 *
 * A pack is a module set: `site/data/<id>/manifest.json` names the content-addressed files that make it up, and
 * `meta` rides on the manifest rather than in any module, because it is the only per-PACK value in an otherwise
 * per-BUILD artifact. Reassembling that into the one flat object {@link SpellPack} describes is the whole job here.
 *
 * It lives apart from the tools that use it because three of them were each doing it, which is three copies of one
 * rule and three chances to disagree — and they did the day the artifact became a module set and nothing noticed.
 *
 * The browser is deliberately NOT a caller. `data.ts` speaks fetch, a relative URL and DecompressionStream, which is
 * its transport rather than its logic; everything that turns bytes into something searchable is shared already. This
 * is the whole of the port: pick the bytes up differently, hand them to the same engine.
 */
import {readFileSync} from "node:fs";
import {fileURLToPath} from "node:url";
import {dirname, resolve} from "node:path";
import {gunzipSync} from "node:zlib";

import type {SpellPack, VersionEntry} from "../src/data";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const SITE = resolve(ROOT, "site");

/** One entry of a pack's manifest: which file holds a module. */
interface ManifestModule {
    file: string;
}

/** What `site/data/<id>/manifest.json` carries. */
interface PackManifest {
    pack: string;
    meta: SpellPack["meta"];
    modules: Record<string, ManifestModule>;
    locales: Record<string, Record<string, ManifestModule>>;
    absentSections?: string[];
}

/**
 * The language a reader gets when it asks for none.
 *
 * Every pack ships it, which is what lets a caller fall back to it without checking. Declared again in
 * `tools/packfile.py` and `build/pack/derive/locales.py`; `check_locale_declaration` reconciles all three.
 */
export const DEFAULT_LOCALE = "enUS";

/** Every version the site ships, newest roster order. */
export function readVersions(): VersionEntry[] {
    return JSON.parse(readFileSync(resolve(SITE, "data", "versions.json"), "utf8")) as VersionEntry[];
}

/** One version's manifest, read off disk. */
function readManifest(entry: VersionEntry): PackManifest {
    return JSON.parse(readFileSync(resolve(SITE, entry.file), "utf8")) as PackManifest;
}

/**
 * The languages a pack ships, in manifest order.
 *
 * What a locale selector may offer for the pack: {@link readPack} serves any of these without falling back, and not
 * every pack ships every language — Epsilon publishes English alone.
 */
export function shippedLocales(entry: VersionEntry): string[] {
    return Object.keys(readManifest(entry).locales);
}

/**
 * The language a pack will actually serve for a request: the asked-for one where shipped, else the default.
 *
 * The lenient half of the policy split — a sweep that must not stop on one pack resolves through this and reports
 * what it got, while a tool measuring one pack refuses instead ({@link import("./dataset").loadPack}).
 */
export function servedLocale(entry: VersionEntry, want: string): string {
    return shippedLocales(entry).includes(want) ? want : DEFAULT_LOCALE;
}

/**
 * Picks one roster entry by a version prefix or a label fragment.
 *
 * @param want A prefix such as `9.2.7`, a label fragment, or nothing for the default pack.
 * @param versions The roster, defaulting to what the site ships.
 * @returns The entry.
 * @throws If nothing matches, naming what there is — a typo'd version is otherwise a silent fall back to the default.
 */
export function pickVersion(want?: string, versions: VersionEntry[] = readVersions()): VersionEntry {
    const entry = want
        ? versions.find((v) => v.id.startsWith(want) || v.label?.includes(want))
        : versions.find((v) => v.default) ?? versions[0];
    if (!entry) throw new Error(`no pack matches "${want}" — have: ${versions.map((v) => v.id).join(", ")}`);
    return entry;
}

/**
 * Reads one version's pack off disk, reassembled from its modules.
 *
 * A section can be SPLIT across modules rather than living in one: `spells` keeps its ids, icons and schools in
 * `core` while its names ride in the locale module, so the two copies are halves of one section and the merge is
 * per COLUMN. Replacing the section wholesale is the obvious wrong answer and it fails loudly on the first read —
 * `spells.ids` simply vanishes.
 *
 * A section that is a bare array or an id-keyed table lives in exactly one module, so it is taken as it comes.
 *
 * @param entry The roster entry to read.
 * @param locale Which language's names and prose to read; a pack that does not carry it falls back to the default.
 * @returns The pack as one object, the shape `buildIndexes` takes.
 */
export function readPack(entry: VersionEntry, locale: string = DEFAULT_LOCALE): SpellPack {
    const manifest = readManifest(entry);
    const spoken = manifest.locales[locale] ?? manifest.locales[DEFAULT_LOCALE] ?? {};
    const pack: Record<string, unknown> = {meta: manifest.meta};
    for (const module of [...Object.values(manifest.modules), ...Object.values(spoken)]) {
        const loaded = JSON.parse(
            gunzipSync(readFileSync(resolve(SITE, module.file))).toString("utf8")) as Record<string, unknown>;
        for (const [section, payload] of Object.entries(loaded)) {
            const held = pack[section];
            pack[section] = isColumns(held) && isColumns(payload)
                ? {...held, ...payload}
                : payload;
        }
    }
    return pack as unknown as SpellPack;
}

/** Whether a payload is a column dict, the one shape two modules can each hold half of. */
function isColumns(value: unknown): value is Record<string, unknown> {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}
